### DISCUSSION TITLE
Deleting folders to put into recycle bin

### DISCUSSION THREAD

**Discussion by ajcaterino** [pnp/pnpcore#1547] [Q&A] [answered]
When deleting a folder using "folder.Delete()", is the folder and its files moved to recycle bin? From my testing it does not appear that it does. Is there a way, that I am missing, to make it do so? If not, is it possible to move a folder to the recycle bin using CreateCopyJobs?

**Comment by jansenbe** [comment]
@ajcaterino :  recycling a folder is possible, but not implemented in PnP Core SDK. I'll create an issue from this discussion to track that ask. If you are in a rush you can use our [custom API approach](https://pnp.github.io/pnpcore/using-the-sdk/basics-customapirequests.html) to implement this yourselves. See https://s-kainet.github.io/sp-rest-explorer/#/entity/SP.Folder for the SharePoint REST APIs for this. Creating a PR for this would even be better :-)

**Reply by ajcaterino** [reply]
I am trying to use the custom API approach, but I am running into a problem, below is a snippet of my code, at this point "folder" is an IFolder object (not a list item) from the following path:  `https://<company>.sharepoint.com/sites/<clientId>/<subSite>/<documentLibrary>/<folder>`. I was assuming that "{List.Id}" would be the Id of the document library at this point, but I am getting a null object reference in the TokenHandler.

```
var apiRequest = new ApiRequest(ApiRequestType.SPORest, "_api/web/lists(guid'{List.Id}')/items({Item.Id})/recycle")
{
	HttpMethod = HttpMethod.Post
};

var apiResponse = folder.ExecuteRequest(apiRequest);
```

In the PnP.Core.Services.TokenHandler class, while executing the method ResolveTokenAsync on line 236 the code calls GetParentDataModel which does not return an IList (it is still returning an IFolder), see snippet:

```
else if (pnpObject is Model.SharePoint.IFolder folder)
{
    listItem = GetParentDataModel(folder as IMetadataExtensible) as Model.SharePoint.IListItem;
    list = GetParentDataModel(listItem as IMetadataExtensible) as Model.SharePoint.IList;
}
```

Since the "listItem" is null, because it can't convert the object to IList, the next call to GetParentDataModel results in a null object error. Should I be calling the "ExecuteRequest" from a different object (i.e. List, Web, etc.)?

I tried the following, and it took 3 passes of the loop to find the IList.

```
IDataModelParent parent = folder;
var cnt = 0;

while ((parent is IFolder || parent is IManageableCollection) && cnt < 5)
{
    parent = parent.Parent;
    cnt++;
}

if (parent is not IList)
{
    throw new ArgumentException("Unable to determine the list of the folder");
}
```

**Reply by jansenbe** [reply, marked as answer]
@ajcaterino : it does not matter from which object you call `ExecuteRequest` when making custom API calls when you're not using tokens in the API call. In your example you're not using the folder recycle method, but you're recycling a listitem. Please use either `_api/Web/Lists(guid'{List.Id}')/Items({Parent.Id})/Folder` or `_api/Web/Lists(guid'{Parent.Id}')/rootFolder` and replace the tokens before making the call
